### DISCUSSION TITLE
[Prettier Plugin] Sorting classes inside variables

### DISCUSSION THREAD

**Discussion by Daiz** [tailwindlabs/tailwindcss#7558] [Ideas]
**EDIT:** I would like to point out to anyone arriving into this post that this issue has been solved - the Prettier plugin now lets you enable sorting for arbitrary attributes, function calls and template literals. [You can read the readme for more instructions.](https://github.com/tailwindlabs/prettier-plugin-tailwindcss#sorting-non-standard-attributes)

Is there any way to make the plugin sort classes inside variables? Consider the following code:

```typescript
const ClassVariable: React.FC = (props) => {
  const klass = "flex flex-col min-h-0";
  return <div className={klass} />;
};

const ClassName: React.FC = (props) => {
  return <div className="flex flex-col min-h-0" />;
};
```

After formatting, this is the result:

```diff
const ClassVariable: React.FC = (props) => {
-  const klass = "flex flex-col min-h-0"; // not changed
  return <div className={klass} />;
};

const ClassName: React.FC = (props) => {
+  return <div className="flex min-h-0 flex-col" />; // changed
};
```

I can imagine trying to sort all string variables by default would not be feasible, but some kind of workaround would be nice. `klass` is a pretty common variable name for CSS classes from what I've seen, so trying to sort strings assigned to `klass` (as well as strings inside an array assigned to `klass`, like `klass = ["some class strings", bool ? "this" : "that"]`) could be one possibility. I imagine this would be one thing that could use a config option if provided though, like a variable name string/regex that is used for matching what variables to try to sort.

**Comment by StErMi** [comment]
I have zero experience on how to build this plugin and what is possible or not but I add my 2 cents:

would it be possible to see which variables are used to form the `className` value and loop until the source and format them? It could be possible that the value is composed by joining multiple variables so you must loop until you have found the source (where no variables are referenced anymore)

```
const ClassVariable: React.FC = (props) => {
    const k1Color = prop.success ? 'text-green-500' : 'text-red-500';
    const k1 = props.small ? `min-h-0 ${k1Color}` : `min-h-full ${k1Color}`;

    const klass = `flex flex-col ${k1} `;
    return <div className={klass} />;
};
```
So the plugin will look at which variables are used for the className value and check `klass`, `k1` and `k1Color` and stop at `k1Color` because in this case those classes do not reference a variable anymore

**Comment by dcastil** [comment]
I also want to suggest a simple solution to implement though it requires a bit more work for the user. There could be a comment which enables sorting in a subsequent string. Something like the `// eslint-disable-next-line` comment.

I already use the inline comment `/*tw:*/` to enable syntax highlighting for Tailwind class strings in variables via the Tailwind VS Code extension with the VS Code setting

```json
{ "tailwindCSS.experimental.classRegex": ["\\/\\*\\s?tw:\\s?\\*\\/\\s?'(.*)'"] }
```

which then looks like this in JS code

```js
const classNameHiddenOnHover = isDatepickerOpen
    ? /*tw:*/ 'opacity-0'
    : /*tw:*/ 'group-hover:opacity-0 group-focus-within:opacity-0'
```

A comment like this could also be used to enable class sorting in arbitrary places.

**Reply by michaelschufi** [reply]
I agree. It would make most sense if we could configure this in the plugin's options. This way, we can exactly control which parts of the code should be treated as a Tailwind class group.

**Comment by gustavopch** [comment]
Leaving a more complex use case here with ternaries and maps to be considered too:
```js
const buttonClass = [
  'block transition',

  `${full ? 'w-full' : 'w-fit'} ${
    {
      small: 'h-9',
      medium: 'h-12',
      large: 'h-14',
    }[size]
  } rounded-full ${
    {
      neutral: 'border-0',
      primary: 'border border-blue-500',
      secondary: 'border border-blue-500',
      text: 'border-0',
    }[variant]
  } ${
    {
      neutral: 'bg-gray-900/5',
      primary: 'bg-blue-500',
      secondary: 'bg-transparent',
      text: 'bg-transparent',
    }[variant]
  } ${
    {
      small: 'py-2 px-3',
      medium: 'py-3 px-6',
      large: 'py-4 px-10',
    }[size]
  }`,
].join(' ')
```
I haven't put a lot of thought into it, but I'm not sure if the `tailwindCSS.experimental.classRegex` approach would work here (unless perhaps if I added the comment to each line, which would not be ideal IMO).

**Reply by eldiegod** [reply]
this is just kinda crazy no? xd

**Reply by Jasperrr91** [reply]
Oof this example is pretty wild. Why use the template string with spaces if you can leverage the `[].join('')`? And why not combine the duplicate `size` and `variant` maps?

You could even use something like `cva` to make things even easier to read (and have the option for easy compoundVariant styling):

```tsx
const buttonClass = cva(['block transition'], {
	variants: {
		full: { true: 'w-full', false: 'w-fit' },
		size: {
			small: ['h-9', 'py-2 px-3'],
			medium: ['h-12', 'py-3 px-6'],
			large: ['h-14', 'py-4 px-10'],
		},
		variant: {
			neutral: ['bg-gray-900/5'],
			primary: ['border border-blue-500', 'bg-blue-500'],
			secondary: ['border border-blue-500', 'bg-transparent'],
			text: ['bg-transparent'],
		},
	},
});
```

**Comment by Daiz** [comment]
@gustavopch The plugin actually already sorts the strings in your example if you put it inside a `<div className={/* complex array/ternary/whatever goes here */}>`. In general, the plugin works by identifying specific AST nodes and sorting all strings that appear inside said nodes. So in theory, all we should need is some way to signal to the plugin when a variable is a *class name variable* and it should be able to handle things from there.

**Comment by liuqiang1357** [comment]
Is it possible to use "Tagged Template Literals" in ES6 to give a signal to this plugin? For example above, maybe we can use it like this:
```const klass = tw`flex flex-col min-h-0`;```

**Reply by chiubaca** [reply]
i scrolled past this, but i got this half working. See my comment - https://github.com/tailwindlabs/tailwindcss/discussions/7558#discussioncomment-4676228

**Comment by cbn-falias** [comment]
Another idea:
Sort classes when they are inside a `classNames()` function call, as this is quite a popular [package](https://www.npmjs.com/package/classnames).
```ts
const klass = classNames("flex flex-col min-h-0");
return <div className={klass} />;
```

Edit: There exists already an idea for that https://github.com/tailwindlabs/tailwindcss/discussions/6382 😉

**Comment by michaeloliverx** [comment]
It would be nice if we could annotate structures to sort. I would like to be able to sort the following:
```js
const baseClasses = 'rounded-md font-medium focus:outline-none'

const variantsLookup = {
  primary: 'bg-cyan-500 text-white shadow-lg hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan-500',
  secondary: 'bg-slate-200 text-slate-800 shadow hover:bg-slate-300 focus:bg-slate-300 focus:ring-slate-500',
  danger: 'bg-red-500 text-white shadow-lg uppercase tracking-wider hover:bg-red-400 focus:bg-red-400 focus:ring-red-500',
  text: 'text-slate-700 uppercase tracking-wider hover:underline hover:text-slate-600 hover:bg-slate-900/5 focus:text-slate-600 focus:ring-slate-500',
}

const sizesLookup = {
  small: 'px-3 py-1.5 text-sm focus:ring-2 focus:ring-offset-1',
  medium: 'px-5 py-3 focus:ring-2 focus:ring-offset-2',
  large: 'px-8 py-4 text-lg focus:ring focus:ring-offset-2',
}

export const Button = (props) => {
  const { variant, size, ...rest } = props
  return <button {...rest} className={`${baseClasses} ${variantsLookup[variant]} ${sizesLookup[size]}`} />
}
```

**Comment by eckoln** [comment]
We're using [Class Variance Authority](https://github.com/joe-bell/cva) package for building tailwind css design system. Any chance sorting for this?

```js
const buttonStyles = cva(
  "flex px-3.5 py-1.5 rounded font-medium focus:outline-none text-white justify-center items-center transition-colors",
  {
    variants: {
      intent: {
        primary: "bg-blue-500 hover:bg-blue-600",
      },
      fullWidth: {
        true: "w-full",
      },
    },
    defaultVariants: {
      intent: "primary",
    },
  }
);

export function Button({ intent, fullWidth, ...props }) {
  return <button className={buttonStyles({ intent, fullWidth })} {...props} />;
}
```

**Reply by kharann** [reply]
> We're using [Class Variance Authority](https://github.com/joe-bell/cva) package for building tailwind css design system. Any chance sorting for this?
> 
> ```js
> const buttonStyles = cva(
>   "flex px-3.5 py-1.5 rounded font-medium focus:outline-none text-white justify-center items-center transition-colors",
>   {
>     variants: {
>       intent: {
>         primary: "bg-blue-500 hover:bg-blue-600",
>       },
>       fullWidth: {
>         true: "w-full",
>       },
>     },
>     defaultVariants: {
>       intent: "primary",
>     },
>   }
> );
> 
> export function Button({ intent, fullWidth, ...props }) {
>   return <button className={buttonStyles({ intent, fullWidth })} {...props} />;
> }
> ```

I also have the same use case! Would love to be able to sort these classes. The experimental regex for autocomplete works however

**Reply by nderscore** [reply]
I recently made a fork of the tailwindcss plugin in order to support nativewind (v3.x), which implements a CVA-esque variants API:

https://github.com/nderscore/prettier-plugin-nativewind

I just added a feature in my fork which lets you customize which props and function calls the plugin searches for tailwind classes in.

With that change, it should be able to be configured to work with `cva` now too.

**Reply by rhkdgns95** [reply]
Change .prettierrc `tailwindFunctions: ["cva", "cn"]`

```json
{
  "tailwindFunctions": [
    "cva",
    "cn"
  ],
  "plugins": ["prettier-plugin-tailwindcss"],
  "endOfLine": "auto"
}

```

**Reply by PaulHaze** [reply]
> Change .prettierrc `tailwindFunctions: ["cva", "cn"]`
> 
> ```json
> {
>   "tailwindFunctions": [
>     "cva",
>     "cn"
>   ],
>   "plugins": ["prettier-plugin-tailwindcss"],
>   "endOfLine": "auto"
> }
> ```

Just tried this approach and so far its not working in a cva-shadcn button...

EDIT: I stand corrected, restarting the program and changing the order and now it works. sorting classes in the start of a cva block:

```
  plugins: ['prettier-plugin-tailwindcss'],
  tailwindFunctions: ['cva', 'cn'],
```

**Comment by bmadone** [comment]
Any updates?

**Comment by eldiegod** [comment]
damn, hoping for this to work someday

**Comment by chiubaca** [comment]
At work we agreed a clean DX for this could be with a tagged literal. This is inspired from writing GROQ queries, for anyone that uses SanityCMS. They provide a [`groq`](https://github.com/sanity-io/sanity/tree/next/packages/groq) tagged literal and when used in conjunction with their [vs code extension](https://marketplace.visualstudio.com/items?itemName=sanity-io.vscode-sanity) it give you syntax highlighting and other niceties. 

So with this DX in mind it would be neat to have something like a `tw` tagged literal  that worked something  like this:

```js
import { tw } from '@tailwindcss'

const tailwindClass = tw`p-10 m-10`;  // get full tw prettier and intellisense here
```

I was able to get this half working by coping how Sanity does this, but prettier sorting doesn't work...

<img width="840" alt="Screenshot 2023-01-13 at 08 45 16" src="https://user-images.githubusercontent.com/18376481/212281675-b6da5f67-5dee-49f3-8c7b-390edd7ce449.png">

**Here are the manual steps to achieve this**

1) Add this file
[based off `groq.ts`](https://github.com/sanity-io/sanity/blob/next/packages/groq/src/groq.ts)
```ts
//my-helpers/tw-tag.ts

export function tw(strings: TemplateStringsArray, ...keys: any[]): string {

  const lastIndex = strings.length - 1;
  return (
    strings.slice(0, lastIndex).reduce((acc, str, i) => {
      return acc + str + keys[i];
    }, '') + strings[lastIndex]
  );
}
```

2) Import the helper in your component file:

```js
import { tw } from '..my-helpers/tw-tag';

const tailwindClass = tw`bg-slate-400 md:bg-slate-400 p-10`;

```

3) Finally, configure the tailwind extension to look inside the `tw`  tag.

```js
{
  "tailwindCSS.experimental.classRegex": [
    "tw`([^`]*)", // tw`...`
  ]
}

```

The final challenge is for tailwind prettier to auto sort the string which I'm not sure how to do...

**Reply by michaeloliverx** [reply]
This looks interesting, I wonder if the tailwind helper simply returned what was passed in then maybe a bundler/minifier  could just inline the function call.

**Reply by drMonro** [reply]
Anybody found way for auto sort in tw files?

**Comment by bettiolo** [comment]
please add reorder in `cva()` 🙏

**Reply by prince272** [reply]
How far?

**Comment by kjrocker** [comment]
I have another use case where I don't _have_ a className variable, because I'm passing classes to a complex third party component.

Messing with `classRegex` gets me intellisense, but prettier isn't picking it up. It would be great if prettier could inspect this setting, I think that solves most of the use-cases people have listed here.

`<ThirdPartyAuthUI appearance={{ className: { button: "", input: "", anchor: "" } }}/>`

**Reply by agusterodin** [reply]
I know the Tailwind maintainers don't want us to do this sort of thing but there are situations where putting classnames in a variable make a lot of sense (as opposed to component abstraction or mapping). For example:

<img width="1027" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/10248395/33ffd1a1-e8f0-4dc5-b7fe-7120f77aa12a">

```
  "tailwindCSS.classAttributes": ["class", "className", "ngClass", ".*className", "cn", "classNames"]
```

Intellisense pick the `noiseClassName` const variable up fine if I set it up to do so with above snippet. Unfortunately there doesn't seem to be a way to do something similar with the Prettier plugin.

**Comment by Daiz** [comment]
I've edited the opening post to say as much, but the Prettier plugin can in fact now achieve what was asked in this request by allowing you to define sorting for arbitrary attributes, function calls and template literals. [You can read the readme for the instructions.](https://github.com/tailwindlabs/prettier-plugin-tailwindcss#sorting-non-standard-attributes)

**Reply by Jasperrr91** [reply]
Sorry but this is not fixed. It works for your initial use case of having `klass` attributes but you need to manually define everything. If for some reason you cannot use the same variable name everytime, like if you have some sort of `InnerContainerClass`/`OuterContainerClass` (for whatever reason) this is going to become a hell to maintain.

Programming wise it might be a bit tedious but the solution is to parse and traverse for the likes of `className=`, note each variable reference or function call -> traverse again and sort those variables, etcetera. This issue should stay open.

**Reply by dcastil** [reply]
You could create a `const tw = String.raw` and then wrap all the Tailwind class strings in `tw<backtick>px-2 flex foo bar<backtick>`. But I'm not sure if I like the approach to add complexity to the runtime just for sorting them.

**Reply by Jasperrr91** [reply]
Yea not ideal but a clever solution nonetheless.

**Reply by Daiz** [reply]
You could also go with `const tw = (str) => str;` and then get sortable tailwind strings with `tw("px-2 flex foo bar")`. I would personally consider that sufficient for my needs, an identity function is not exactly much overhead for something like this.

**Reply by michaelschufi** [reply]
This breaks if you want to use CVA, no? Meaning, you always have to have a function or some other wrapper to make it work.
I think having the option to mark an arbitrary object value, array, JSX attribute, etc. as "tailwind class context" is needed.

**Reply by nderscore** [reply]
> This breaks if you want to use CVA, no? Meaning, you always have to have a function or some other wrapper to make it work.

CVA is already a function. No additional wrapper is needed. Just add `cva` to the `tailwindFunctions` config option.

> I think having the option to mark an arbitrary ... JSX attribute ... is needed.

This is already supported via the `tailwindAttributes` config option.

**Reply by aleczratiu** [reply]
On my side the sorting for cva is not working even if is added in tailwindFunctions.

```
"tailwindFunctions": ["clsx", "cn", "twmerge", "cva"]
```

Any updates here?

**Reply by loczek** [reply]
Make sure to add the `prettier-plugin-tailwindcss` plugin as well

```json
"plugins": ["prettier-plugin-tailwindcss"],
"tailwindFunctions": ["clsx", "cn", "twmerge", "cva"]
```

**Reply by aleczratiu** [reply]
![Screenshot 2023-10-22 at 8 30 50 AM](https://github.com/tailwindlabs/tailwindcss/assets/33958032/4bb1a13f-1fd0-49f9-9342-22fabd3417ae)

Indeed, is like that and is not formatting it and it still seems to be with warnings on the file.

![Screenshot 2023-10-22 at 8 33 30 AM](https://github.com/tailwindlabs/tailwindcss/assets/33958032/a4cd3453-9887-4d73-9f5c-8a38bec79bab)

**Reply by marckraw** [reply]
any update on that ? It doesnt work for me aswell. We are trying to sort classes in nested `cva` object

**Reply by zrtsky** [reply]
final `.prettierrc` file needs to look like this:

```json
{
  "plugins": [
    "prettier-plugin-tailwindcss"
  ],
  "tailwindFunctions": ["cva"]
}
```

**Reply by QingjiaTsang** [reply]
any updates?  my file looks like this and it's not working when it comes to sorting classnames in the cn function
```
{
  "bracketSpacing": true,
  "endOfLine": "lf",
  "plugins": ["prettier-plugin-tailwindcss"],
  "printWidth": 100,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 2,
  "tailwindFunctions": ["clsx", "twMerge", "cn"],
  "trailingComma": "all"
}
```

**Reply by gustavopch** [reply]
It should be working, so there's something else going on. I suggest you create a repro on Stackblitz or similar and paste the link here.

**Comment by ravshansbox** [comment]
How can we sort classes in projects which uses [windstitch](https://windstitch.vercel.app/)?

**Comment by Psycarlo** [comment]
How would I be able to auto-sort this one?

```typescript
const buttonVariants = {
  variant: {
    default: 'bg-gray-600',
    white: 'bg-white',
    ghost: 'bg-blue-500'
  },
  text: {
    default: 'text-white uppercase',
    white: 'text-black uppercase',
    ghost: 'uppercase'
  }
}

const buttonStyles = cva(['h-14 flex-row justify-center items-center w-full'], {
  variants: buttonVariants,
  defaultVariants: {
    variant: 'default',
    text: 'default'
  }
})
```

**Comment by BoilingSoup** [comment]
Does `tailwindFunctions` only work for javascript/typescript?

I'm using the `experimental.classRegex` tailwind setting to get autocomplete in `.go` files.
I then tried adding `tailwindFunctions` in my `.prettierrc` but it's not sorting. 
```
{
    "plugins": [
        "prettier-plugin-tailwindcss"
    ],
    "tailwindFunctions": [
        "Class",
        "Classes"
    ]
}
```

The autocomplete works good, it just doesn't sort.
